tests/feature_resolution_editions.js

```diff
@@ -0,0 +1,300 @@
+
+    string string_val = 1;
+    string string_repeated = 2 [features.amazing_feature = F];
+
+    uint64 uint64_val = 3;
+    uint64 uint64_repeated = 4;
+
+    bytes bytes_val = 5;
+    bytes bytes_repeated = 6;
+
+    SomeEnum enum_val = 7;
+    SomeEnum enum_repeated = 8;
+
+    extensions 10 to 100;
+    extend Message {
+        int32 bar = 10 [features.amazing_feature = I];
+    }
+
+    message Nested {
+        option features.amazing_feature = H;
+        int64 count = 9;
+    }
+
+    enum SomeEnumInMessage {
+        option features.amazing_feature = G;
+        ONE = 11;
+        TWO = 12;
+    }
+    
+    oneof SomeOneOf {
+        option features.amazing_feature = J;
+        int32 a = 13;
+        string b = 14;
+    }
+
+    map<string,int64> int64_map = 15;
+}
+
+extend Message {
+    int32 bar = 16 [features.amazing_feature = D];
+}
+
+enum SomeEnum {
+    option features.amazing_feature = C;
+    ONE = 1 [features.amazing_feature = K];
+    TWO = 2;
+}
+*/
+var tape = require("tape");
+
+var protobuf = require("..");
+
+var protoEditions2023 = `edition = "2023";`;
+
+var proto2 = `syntax = "proto2";`;
+
+var proto3 = `syntax = "proto3";`;
+
+var editions2023Defaults = {enum_type: 'OPEN', field_presence: 'EXPLICIT', json_format: 'ALLOW', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'PACKED', utf8_validation: 'VERIFY'}
+var proto2Defaults = {enum_type: 'CLOSED', field_presence: 'EXPLICIT', json_format: 'LEGACY_BEST_EFFORT', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'EXPANDED', utf8_validation: 'NONE'}
+var proto3Defaults = {enum_type: 'OPEN', field_presence: 'IMPLICIT', json_format: 'ALLOW', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'PACKED', utf8_validation: 'VERIFY'}
+
+tape.test("feature resolution defaults", function(test) {
+    var rootEditions = protobuf.parse(protoEditions2023).root;
+    rootEditions.resolveAll();
+    test.same(rootEditions._features, editions2023Defaults);
+
+    var rootProto2 = protobuf.parse(proto2).root;
+    rootProto2.resolveAll();
+    test.same(rootProto2._features, proto2Defaults);
+
+    var rootProto3 = protobuf.parse(proto3).root;
+    rootProto3.resolveAll();
+    test.same(rootProto3._features, proto3Defaults);
+
+    test.end();
+})
+
+tape.test("feature resolution inheritance", function(test) {
+    var rootEditionsOverriden = protobuf.parse(`edition = "2023";
+    option features.json_format = LEGACY_BEST_EFFORT;
+
+    option features.(abc).d_e = deeply_nested_false;
+
+    message Message {
+        string string_val = 1;
+        string string_repeated = 2 [features.enum_type = CLOSED];
+
+        message Nested {
+            option features.(abc).d_e = deeply_nested_true;
+            option features.field_presence = IMPLICIT;
+            int64 count = 9;
+        }
+        
+        enum SomeEnum {
+            ONE = 1 [features.repeated_field_encoding = EXPANDED];
+            TWO = 2;
+        }
+    }`).root
```

Comment:
I think this is missing some cases, but it might be useful to split it up even further.  For example, a test covering file->message inheritance in isolation, instead of trying to bundle them all in 1 monolithic test.  The problem with this pattern is that it's still really difficult to verify the behavior by inspection.

Missing cases: oneof, extensions, top-level enum, service, method

Reply:
Added tests.